h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -160,6 +173,124 @@ public long getModelCount() {
+      // Attempt to train next model
+      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+      if (nextModelParams != null && isThereEnoughTime() && !_job.stop_requested()) {
+        parallelModelBuilder.run(ModelBuilder.make(nextModelParams));
+      } else {
+        parallelModelBuilder.noMoreModels();
+      }
+    }
+
+    private void constructScoringInfo(final Model model) {
+      ScoringInfo scoringInfo = new ScoringInfo();
+      scoringInfo.time_stamp_ms = System.currentTimeMillis();
+
+      model.fillScoringInfo(scoringInfo);
+      grid.setScoringInfos(ScoringInfo.prependScoringInfo(scoringInfo, grid.getScoringInfos()));
+      ScoringInfo.sort(grid.getScoringInfos(), _hyperSpaceWalker.search_criteria().stopping_metric());
+    }
+
+    private boolean isThereEnoughTime() {
+      final boolean enoughTime = hyperspaceIterator.max_runtime_secs() > 0 && hyperspaceIterator.time_remaining_secs() > 0;
+
+      if (!enoughTime) {
+        Log.info("Grid max_runtime_secs of " + hyperspaceIterator.max_runtime_secs() + " secs has expired; stopping early.");
+      }
+
+      return enoughTime;
+    }
+
+    private MP getNextModelParams(final HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceWalker, final Model model, final Grid grid){
+      MP params = null;
+
+      while (params == null) {
+        if (hyperSpaceWalker.hasNext(Optional.of(model))) {
+          params = hyperSpaceWalker.nextModelParameters(Optional.ofNullable(model));
+          final Key modelKey = grid.getModelKey(params.checksum(IGNORED_FIELDS_PARAM_HASH));
+          if(modelKey != null){
+            params = null;
+          }
+        } else {
+          break;
+        }
+      }
+
+      return params;
+    }
+  }
+
+  /**
+   * Searches the hyperspace and builds models in a parallel way - building the models in parallel.
+   *
+   * @param grid Grid to add models to
+   */
+  private void parallelGridSearch(final Grid<MP> grid) {
+    final HyperSpaceWalker.HyperSpaceIterator<MP> iterator = _hyperSpaceWalker.iterator();
+    final ModelFeeder modelFeeder = new ModelFeeder(iterator, grid);
+    final ParallelModelBuilder parallelModelBuilder = new ParallelModelBuilder(modelFeeder::onModelSuccessfullyBuilt,
+            modelFeeder::onModelBuildFailed);
+
+    List<ModelBuilder> startModels = new ArrayList<>();
+    final int parallelismLevel = _parallelismLevel > 1 ? _parallelismLevel : 2 * H2O.NUMCPUS;
+    final List<MP> mps = iterator.initialModelParameters(parallelismLevel);
+
+    for (int i = 0; i < mps.size(); i++) {
+      final MP nextModelParameters = mps.get(i);
+      final long checksum = nextModelParameters.checksum(IGNORED_FIELDS_PARAM_HASH);
+      if (grid.getModelKey(checksum) == null) {
+        startModels.add(ModelBuilder.make(nextModelParameters));
+      } else {
+        //TODO: Compensate for this scenario - we need to keep the level of paralelism at pre-set level
+      }
+    }
```

Comment:
not a big fan of this approach.
I think you could use an ArrayBlockingQueue limited to parallelism, and fill it continuously as soon as it has capacity (waiting when it's full) and as soon as you have next model to train.